application/client/src/app/ui/views/toolbar/chart/zoomer/state.ts

```diff
@@ -0,0 +1,233 @@
+    }
+
+    public getLeftOffset(): number {
+        if (this._filters === undefined) {
+            return 0;
+        }
+        return this._filters.chartArea.left;
+    }
+
+    private _subscribe(containerWidth: number) {
+        if (this._env === undefined || this._ilc === undefined) {
+            this._logger.warn("Couldn't subscribe to chart events.");
+            return;
+        }
+        this._env.subscriber.register(
+            this._session.search.subjects.get().updated.subscribe(this._update.bind(this)),
+        );
+        this._env.subscriber.register(
+            this._session.search
+                .store()
+                .filters()
+                .subjects.get()
+                .highlights.subscribe((entity: StoredEntity<FilterRequest>) => {
+                    this._changeColor(entity.definition.uuid, entity.definition.colors.background);
+                }),
+        );
+        this._env.subscriber.register(
+            this._ilc.channel.ui.sidebar.resize(this._resize(containerWidth)._update.bind(this)),
+        );
+    }
+
+    private _update() {
+        if (this._filters !== undefined && this._filters.width > 0) {
+            this._fetch(this._filters.width);
+        }
+    }
+
+    private _changeColor(uuid: string, backgroundColor: string) {
+        if (this._filters !== undefined) {
+            this._filters.data.datasets.forEach((dataset) => {
+                if (dataset.label === uuid) {
+                    dataset.backgroundColor = backgroundColor;
+                    return;
+                }
+            });
+            this._filters.update();
+        }
+    }
+
+    private _resize(containerWidth?: number): State {
+        if (containerWidth) {
+            this._labels = new Array(containerWidth).fill('');
+        }
+        this._detectChanges();
+        return this;
+    }
+
+    // [TODO] Use this function to show or hide cursor
+    private _updateCursor(): boolean {
+        const prev: boolean = this.isCursorVisible;
+        // const streamSize: number | undefined = this.serviceData.getStreamSize();
+        // if (streamSize === undefined || !this.serviceData.hasData()) {
+        //     this.isCursorVisible = false;
+        // } else if (this.width < streamSize) {
+        //     this.isCursorVisible = true;
+        // } else {
+        //     this.isCursorVisible = false;
+        // }
+        if (this.isCursorVisible !== prev) {
+            this._detectChanges();
+            return true;
+        }
+        return false;
+    }
+
+    private _draw() {
+        this._loading = true;
+        this._extractData();
+        if (this._filters === undefined) {
+            this._createChart();
+        } else {
+            this._filters.data = {
+                datasets: this._datasets,
+                labels: this._labels,
+            };
+            this._filters.update();
+        }
+        this._loading = false;
+        this._detectChanges();
```

Comment:
this piece of code looks as synchronic code: started at line 136 and finished on 148. It isn't clear why you need flag `this._loading` here? It make sense if you are doing some async stuff. I'm confused a bit with it